common/src/test/java/com/revenuecat/purchases/common/networking/MapConverterTest.kt

```diff
@@ -0,0 +1,91 @@
+package com.revenuecat.purchases.common.networking
+
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import io.mockk.every
+import io.mockk.mockk
+import io.mockk.spyk
+import org.json.JSONArray
+import org.json.JSONObject
+import org.junit.Assert.assertEquals
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.annotation.Config
+
+@RunWith(AndroidJUnit4::class)
+@Config(manifest = Config.NONE)
+class MapConverterTest {
+
+    private lateinit var mapConverter: MapConverter
+
+    @Before
+    fun setUp() {
+        mapConverter = MapConverter()
+    }
+
+    @Test
+    fun `test convert to JSON`() {
+        val inputMap = mapOf(
+            "key1" to "value1",
+            "key2" to listOf("value2", "value3"),
+            "key3" to mapOf("nestedKey" to "nestedValue")
+        )
+
+        val expectedJson = JSONObject()
+            .put("key1", "value1")
+            .put("key2", JSONArray(listOf("value2", "value3")))
+            .put("key3", JSONObject().put("nestedKey", "nestedValue"))
+
+        val result = mapConverter.convertToJSON(inputMap)
+        assertEquals(expectedJson.toString(), result.toString())
+    }
+
+    @Test
+    fun `test convert to JSON with nested array of strings`() {
+        val inputMap = mapOf(
+            "key1" to "value1",
+            "key2" to listOf("value2", "value3"),
+            "key3" to mapOf("nestedKey" to "nestedValue"),
+            "key4" to listOf("value4", "value5")
+        )
+
+        val expectedJson = JSONObject()
+            .put("key1", "value1")
+            .put("key2", JSONArray(listOf("value2", "value3")))
+            .put("key3", JSONObject().put("nestedKey", "nestedValue"))
+            .put("key4", JSONArray(listOf("value4", "value5")))
+
+        val result = mapConverter.convertToJSON(inputMap)
+        assertEquals(expectedJson.toString(), result.toString())
+    }
+
+    /**
+     * This tests workaround for a bug in Android 4 , where a List<String> would be incorrectly converted into
+     * a single string instead of a JSONArray of strings.
+     * (i.e.: "[\"value1\", \"value2\"]" instead of "[value1, value2]")
+     */
+    @Test
+    fun `test map conversion fixes wrong treatment of arrays of strings in JSON library`() {
+        val mapConverterMock = spyk<MapConverter>()
```

Comment:
Maybe rename to `mapConverterSpy`? Since it's not technically a mock

Reply:
ohh spyk nice, I've never used this

Reply:
I'm not technically using it as a spy either, more like a partial mock, since I do modify the return value for _some_ circumstances. Perhaps renaming it to that would be clearer?

Reply:
trying it this way, let me know what you think